orbit-android/src/main/java/com/babylon/orbit/OrbitViewModel.kt

```diff
@@ -51,16 +56,35 @@ open class OrbitViewModel<STATE : Any, SIDE_EFFECT : Any>(
         stateConsumer: (STATE) -> Unit,
         sideEffectConsumer: (SIDE_EFFECT) -> Unit = {}
     ) {
-        mainThreadOrbit
-            .autoDispose(lifecycleOwner)
-            .subscribe(stateConsumer)
+        disposables += mainThreadOrbit.subscribe(stateConsumer)
+        disposables += mainThreadSideEffect.subscribe(sideEffectConsumer)
 
-        mainThreadSideEffect
-            .autoDispose(lifecycleOwner)
-            .subscribe(sideEffectConsumer)
+        val exitEvent = lifecycleOwner.lifecycle.currentState.correspondingExitEvent
+        lifecycleOwner.lifecycle.addObserver(object : LifecycleEventObserver {
+            override fun onStateChanged(source: LifecycleOwner, event: Lifecycle.Event) {
+                if (event == exitEvent) {
+                    disposables.clear()
+                    source.lifecycle.removeObserver(this)
+                }
+            }
+        })
     }
 
     override fun onCleared() {
+        disposables.dispose()
         container.disposeOrbit()
     }
+
+    private val Lifecycle.State.correspondingExitEvent: Lifecycle.Event
+        get() = when (this) {
+            Lifecycle.State.DESTROYED -> error("Lifecycle is already destroyed")
+            // in onCreate
+            Lifecycle.State.INITIALIZED -> Lifecycle.Event.ON_DESTROY
+            // in onStart
+            Lifecycle.State.CREATED -> Lifecycle.Event.ON_STOP
+            // in onResume
+            Lifecycle.State.STARTED -> Lifecycle.Event.ON_PAUSE
+            // between onResume and onPause
+            Lifecycle.State.RESUMED -> Lifecycle.Event.ON_PAUSE
```

Comment:
I found this really confusing. In `Activity.onStart` `currentState` is `CREATED` but if I register an observer then when it receives `ON_START` `currentState` will be `STARTED`.

Reply:
Yep, it's pretty unintuitive. The `Lifecycle` class javadoc sheds some light on this behaviour:

> Lifecycle.Event.ON_CREATE, Lifecycle.Event.ON_START, Lifecycle.Event.ON_RESUME events in this class are dispatched after the LifecycleOwner's related method returns. Lifecycle.Event.ON_PAUSE, Lifecycle.Event.ON_STOP, Lifecycle.Event.ON_DESTROY events in this class are dispatched before the LifecycleOwner's related method is called. For instance, Lifecycle.Event.ON_START will be dispatched after onStart returns, Lifecycle.Event.ON_STOP will be dispatched before onStop is called. This gives you certain guarantees on which state the owner is in.

As far as I know `autodispose` registers an observer instead, interesting there are a few ways to do this.